Lighten and test constraints on context types in alternative implementations

gilles-peskine-arm opened the pull request:
* Write some basic documentation about alt implementation. This wasn't really in scope, but I needed some place to write the documentation for this task. Writing detailed, polished documentation is out of scope of this pull request.
* Fix some glitches of the ECP alt interface.
* Remove `MBEDTLS_ECP_MAX_BITS` as a configuration option.
* Add some accessor functions to the DHM and ECJPAKE modules to relax the constraints on context types from `MBEDTLS_DHM_ALT` and `MBEDTLS_ECJPAKE_ALT`.
* Tweak the code to relax constraints on the context types of several whole-module alternative implementations.
* Test the build with whole-module `MBEDTLS_XXX_ALT` defined. (Just building the library. Anything involving linking is out of scope.)
* Document the remaining constraints on context types.
* Add the constraint that context types must be movable. Fix #4451.

Fix https://github.com/ARMmbed/mbedtls/issues/4573.

Size note: it took me about a day to assemble this PR, so it fits in S. However, due to the diversity of the changes, this will be somewhat lengthy to review, hence estimating at M for review.

mpg commented on tests/scripts/all.sh:
```diff
@@ -1383,6 +1383,21 @@ component_build_default_make_gcc_and_cxx () {
+component_build_module_alt () {
+    msg "build: MBEDTLS_XXX_ALT" # ~30s
+    scripts/config.py full
+    # Disable options that are incompatible with some ALT implementation.
```
Perhaps we should have comments explaining the incompatibility. Debug is not obvious to me, and it's not obvious to me either why AESNI is incompatible while PADLOCK isn't.

gilles-peskine-arm replied:
I determined this list by enabling everything and then disabling the minimum I could to get the library to compile. The code of `padlock.c` is only actually enabled on 32-bit x86, and this component does a default-architecture build which in practice means 64-bit x86. I'll exclude `MBEDTLS_PADLOCK_C` anyway for robustness.

mpg commented on tests/scripts/all.sh:
```diff
@@ -1390,6 +1390,7 @@ component_build_module_alt () {
     scripts/config.py unset MBEDTLS_AESNI_C
     scripts/config.py unset MBEDTLS_DEBUG_C
     scripts/config.py unset MBEDTLS_THREADING_PTHREAD
+    scripts/config.py unset MBEDTLS_PK_PARSE_EC_EXTENDED
```
Actually, shouldn't we have a `check_config.h` check for that (and the others above)?

Edit: probably not, the limitations are documented in the next commit and a more subtle than what `check_config.h` can handle. OK.

mpg reviewed (changes requested):
Looks pretty good to me, just a couple minor suggestions for improvements. Thanks for adding these tests and providing some documentation here!

gilles-peskine-arm commented on docs/architecture/alternative-implementations.md:
```diff
@@ -0,0 +1,66 @@
+
+For timing functions, you can [declare an alternative implementation of the timing module](#module-alternative-implementations).
+
+For networking, on systems without a BSD-like socket interface, you can replace `net_sockets.c` by a compatible module, or use custom functions for TLS.
```
You can replace `net_sockets.c` after copying the library sources. But since we don't officially support that, I won't mention it.

gilles-peskine-arm commented on docs/architecture/alternative-implementations.md:
```diff
@@ -0,0 +1,66 @@
+
+### Alternative implementations of platform functions
+
+Several platform functions can be reconfigured dynamically by following the process described here. To reconfigure how Mbed TLS calls the standard library function `xxx()`:
```
I added some examples, which show that the generic pattern I describe is barely followed. The platform interface is pretty messy and I consider a detailed description out of scope.

gilles-peskine-arm commented on include/mbedtls/ecp.h:
```diff
@@ -292,6 +285,44 @@ mbedtls_ecp_group;
+/* Dummy definition to help code that has optional ECP support and
+ * defines an MBEDTLS_ECP_MAX_BYTES-sized array unconditionally. */
+#define MBEDTLS_ECP_MAX_BITS 1
+/* Note: the curves must be listed in DECREASING size! */
```
The canonical list of things to do when adding a curve is above the definition of `mbedtls_ecp_group_id`.

gilles-peskine-arm commented on tests/include/alt-dummy/arc4_alt.h:
```diff
@@ -0,0 +1,30 @@
+/* arc4_alt.h with dummy types for MBEDTLS_ARC4_ALT */
```
When the review here has settled, if #4588 isn't merged yet, I'll rewrite the history to skip modules that are about to be removed in “Create xxx_alt.h headers for testing”.

gilles-peskine-arm replied:
I don't think there will be any merge conflict since this PR doesn't touch places where modules having ALT are referenced. Nor CI failures since this PR doesn't rely on the existence of ALT for modules other than DHM, ECP and threading. Just some useless files which we can remove in a follow-up.

mpg replied:
Agreed, so I'm marking this conversation as resolved.

mpg reviewed (approved):
Thanks for addressing my feedback. Looks all good to me now.

The only point that remains is about the modules that are about to me removed. I think it's nice of you to suggest rewriting your history here, but considering this is a 3.0-mandatory task while module removal is 3.0-optional and seems less advanced in terms of review, I think it's also OK to leave your PR as it is, and then let #4588 handle it.

mpg commented:
Hmm, just noticed there are conflicts.

gilles-peskine-arm commented:
The DHM test code changes conflict with those in #4611. I'll rebase once this PR isn't undergoing active review.

ronald-cron-arm commented on docs/architecture/alternative-implementations.md:
```diff
@@ -0,0 +1,91 @@
+
+For networking, the `net_sockets` module does not currently support alternative implementations. If this module does not work on your platform, disable `MBEDTLS_NET_C` and use custom functions for TLS.
+
+If your platform has a cryptographic accelerator, you can use it via a [PSA driver](#psa-cryptography-drivers) or an [declare an alternative implementation of the corresponding module(s)](#module-alternative-implementations) or [specific functions](#function-alternative-implementations). PSA drivers will ultimately replace the alternative implementation mechanism, but alternative implementation w ...
```
```suggestion
If your platform has a cryptographic accelerator, you can use it via a [PSA driver](#psa-cryptography-drivers) or [declare an alternative implementation of the corresponding module(s)](#module-alternative-implementations) or [specific functions](#function-alternative-implementations). PSA drivers will ultimately replace the alternative implementation mechanism, but alternative implementation will remain supported in at least all Mbed TLS versions of the form 3.x. The interface of PSA drivers is currently still experimental and subject to change.
```

ronald-cron-arm commented on docs/architecture/alternative-implementations.md:
```diff
@@ -0,0 +1,91 @@
+
+## Module alternative implementations
+
+You can replace the code of some modules of Mbed TLS at compile time by a custom implementation of this module. This is possible for low-level cryptography modules (symmetric algorithms, DHM, RSA, ECP, ECJPAKE) and for some platform-related modules (threading, timing). Such custom implementations are called “alternative implementations”, or “ALT implementations” for short.
```
```suggestion
You can replace the code of some modules of Mbed TLS at compile time by a custom implementation of them. This is possible for low-level cryptography modules (symmetric algorithms, DHM, RSA, ECP, ECJPAKE) and for some platform-related modules (threading, timing). Such custom implementations are called “alternative implementations”, or “ALT implementations” for short.
```
Not sure but in case ...

ronald-cron-arm commented on docs/architecture/alternative-implementations.md:
```diff
@@ -0,0 +1,91 @@
+Where a context type needs to have a certain field, the field must have the same type and semantics as in the built-in implementation, but does not need to be at the same position in the structure. Furthermore, unless otherwise indicated, only read access is necessary: the field can be `const`, and modifications to it do not need to be supported. For example, if an alternative implementation of a ...
+
+* AES: if `MBEDTLS_AESNI_C` or `MBEDTLS_PADLOCK_C` is enabled, `mbedtls_aes_context` must have the fields `nr` and `rk`.
+* DHM: if `MBEDTLS_DEBUG_C` is enabled, `mbedtls_dhm_context` must have the fields `P`, ``Q`, `G`, `GX`, `GY` and `K`.
```
```suggestion
* DHM: if `MBEDTLS_DEBUG_C` is enabled, `mbedtls_dhm_context` must have the fields `P`, `Q`, `G`, `GX`, `GY` and `K`.
```

ronald-cron-arm reviewed (changes requested):
This looks almost good to me, I only have a few nitpicks. Thanks for all the additional documentation on alternative implementations, things about alternative implementations are clearer to me now.

gilles-peskine-arm commented on include/mbedtls/dhm.h:
```diff
@@ -301,6 +312,20 @@ size_t mbedtls_dhm_get_bitlen( const mbedtls_dhm_context *ctx );
+ * \return         #MBEDTLS_ERR_DHM_BAD_INPUT_DATA if \p field is invalid.
+ * \return         An \c MBEDTLS_ERR_MPI_XXX error code if the copy fails.
+ */
+int mbedtls_dhm_get_value( mbedtls_mpi *dest,
```
Mbed TLS is unfortunately inconsistent. DHM had no functions with both. RSA puts the input first, but bignum, ECP and ECDH put the output first. On the one hand, I think ECDH is the closest thing to DHM so DHM should follow ECDH. On the other hand, I think that if we consider this function to be a getter rather than a copier (which makes sense since it's called `_get_` and not `_copy_`), it is more natural to view the parameters as (context, selector, destination), which go in this order because we put the context first. So ok, I'll change the order.

ronald-cron-arm reviewed (approved):
Thanks for the changes, LGTM.

mpg reviewed (approved):
Reviewed the latest commits and I'm happy with them too. Unfortunately there's a conflict now.

gilles-peskine-arm commented:
I rebased on top of `development` to handle conflicts due to adjacent changes in `test_suite_dhm.function` and to the introduction of `MBEDTLS_PRIVATE`. The previous version is in https://github.com/gilles-peskine-arm/mbedtls/tree/alt-dummy-headers-3.0-3 and the new version is in https://github.com/gilles-peskine-arm/mbedtls/tree/alt-dummy-headers-3.0-4.

```
git range-diff upstream-public/development alt-dummy-headers-3.0-3 alt-dummy-headers-3.0-4
```
gives reasonable output.

Notes about the rebase:

* "Create xxx_alt.h headers for testing" is still the result of the script in the commit message.
* "DHM: new functions to query the length of the modulus" had a spurious addition of a comment start marker in `dhm.h` which actually belonged to the next commit. I fixed that, and this commit now builds in the default configuration. Other than that, in several places, `development` changed `dhm.len` to `dhm.MBEDTLS_PRIVATE(len)`, and this commit changes those to `mbedtls_dhm_get_len( &dhm )` (or variations using a local variable).
* "Document public fields of mbedtls_ecp_group" changed a bit because it's now making the fields public (by not using `MBEDTLS_PRIVATE` in their definition) and not just documenting them. The commit now removes a few `MBEDTLS_PRIVATE` calls in some sample programs.

I will push a second rebase with an identical end result, squashing a couple of commits that were fixups for trivial mistakes found during review.

ronald-cron-arm reviewed (approved):
I've been on my side through the rebasing of https://github.com/ARMmbed/mbedtls/commit/107de3bbd9e714ae94cece61c7025a622d5d81ca on top of `development(823f594db5acafba98d461adacb68d3331adf8f3)` and ended-up with the same tree as e5702489f503f1fc26debc846cb3f4c0ab0b5b66 but for one change in programs/pkey/ecdsa.c and one change in programs/pke/gen_key.c that are mentioned in https://github.com/ARMmbed/mbedtls/pull/4595#issuecomment-861018831 which look good to me.

mpg reviewed (approved):
I verified the rebase and the history clean-up and I'm happy as well.
